vcd/resource_vcd_catalog.go

```diff
@@ -182,6 +212,14 @@ func genericResourceVcdCatalogRead(d *schema.ResourceData, meta interface{}) err
 		return fmt.Errorf("error retrieving catalog %s : %s", d.Id(), err)
 	}
 
+	// catalog user view is retrieved to get the number of vApp templates and medias
+	catalog, err := adminOrg.GetCatalogById(d.Id(), false)
+	if err != nil {
+		log.Printf("[DEBUG] Unable to find catalog. Removing from tfstate")
+		d.SetId("")
+		return fmt.Errorf("error retrieving catalog %s : %s", d.Id(), err)
```

Comment:
As you are working on this code there is a good opportunity for a fix here and in the  `adminCatalog, err := adminOrg.GetAdminCatalogByNameOrId(d.Id(), false)` error handling a few lines above.

In our provider there are a few historical quirks with how it behaves when a resource or data source is not found (mainly in older resources). The rule of thumb should be:
* **Resource not found** during a read - we must do a `d.SetId("")` and `return nil` (not an error). The reason for that is that if a resource has dissapeared (maybe someone delete it via UI), the consumer will just get an offer to create a new one during `terraform plan`. 
 
Now ideally - we'd like to distinguish between a "resource not found" error and any other error (API timeout, etc). For that reason we have a special [sentinel error](https://earthly.dev/blog/golang-errors/#defining-sentinel-errors) `govcd.ErrorEntityNotFound`  and also two helper functions `govcd.IsNotFound` and `govcd.ContainsNotFound`. Whenever we do SDK code, we aim to provide the `govcd.ErrorEntityNotFound` in the returned error when a resource is not found. You need to _double check_ that `adminOrg.GetAdminCatalogByNameOrId` and `adminOrg.GetCatalogById` both return that `govcd.ErrorEntityNotFound` error when needed.
  
* **Data source not found** during a read - error must be returned to user as that would break downstream operations anyway.

Below is one example how we handle it for best user experience.

Excerpt from resource `read` for NSX-T Edge Gateway. See how it conditionally checks for `govcd.ContainsNotFound(err)` to check if the resource is not found - then it resets the ID and returns nil. But if the error is anything else - it will return it to user as we don't want to remove the resource from statefile if for example a connection timeout occured during terraform run.
```go
	edge, err := org.GetNsxtEdgeGatewayById(d.Id())
	if err != nil {
		if govcd.ContainsNotFound(err) {
			d.SetId("")
			return nil
		}
		return diag.Errorf("could not retrieve NSX-T Edge Gateway: %s", err)
	}
```

Excerpt for data source code (a lot simpler as you can see):
```go
        edge, err := vdc.GetNsxtEdgeGatewayByName(d.Get("name").(string))
	if err != nil {
		return diag.FromErr(fmt.Errorf("could not retrieve NSX-T edge gateway: %s", err))
	}
```

Reply:
Repeated comment.